### PR TITLE
feat: Add repoLink to POST Schema

### PR DESCRIPTION
**Pull request by Ansaar-Syed**
## Pre-Requisites
- [x] Yes, I updated [Authors.md](../blob/main/Authors.md) **OR** this is not my first contribution
- [x] Yes, I included and/or modified tests to cover relevent code **OR** my change is non-technical
- [x] Yes, I wrote this code entirely myself **OR** I properly attributed these changes in [Third Party Notices](../blob/main/THIRD-PARTY-NOTICES.txt)

### CODE REVIEW

File: packages/api/src/api/project/post.ts [hunk 1]
```diff
@@ -25,6 +26,10 @@ export const post = async (req: Request, res: Response) => {
 
       project = new Project(data);
       project.contributors.add(lockedUser);
+      const hostRes = await axios.get(data.repoLink);
```
Comment (SpencerKaiser) [inline packages/api/src/api/project/post.ts:29]: This is unrelated to the line above so let's break it up a bit
```suggestion

      const hostRes = await axios.get(data.repoLink);
```

Also what is `hostRes`? Are you fetching a host? – please rename to something that's a bit more descriptive like `repoFetchRes`

Finally you need a `try ... catch` around this fetch

File: packages/api/src/api/project/post.ts [hunk 2]
```diff
@@ -25,6 +26,10 @@ export const post = async (req: Request, res: Response) => {
...
+      if (hostRes.status !== 200) {
+        throw new Error('oh no repoLink cannot be found');
```
Comment (SpencerKaiser) [inline packages/api/src/api/project/post.ts:31]: This will simply break the request and cause a timeout on behalf of the client; this needs to be a structured response to the front end with a meaningful error

```suggestion
        res.status(400).send('Repo URL could not be validated');
        return;
```

File: packages/database/src/entities/Project.ts [hunk 3]
```diff
@@ -23,6 +23,9 @@ export class Project extends Node<Project> {
   @Property({ columnType: 'text' })
   name: string;
 
+  @Property({ columnType: 'text', nullable: true })
```
Comment (SpencerKaiser) [inline packages/database/src/entities/Project.ts:26]: AC didn't say this was supposed to be optional 😬 Your schema also has it as required

```suggestion
  @Property({ columnType: 'text' })
```

File: packages/database/src/migrations/0005-add-repolink-to-project.ts
Comment (SpencerKaiser) [inline packages/database/src/migrations/0005-add-repolink-to-project.ts:1]: You'll have to recreate the migration after the changes above

File: packages/database/src/entities/Project.ts [hunk 4]
```diff
@@ -23,6 +23,9 @@ export class Project extends Node<Project> {
...
+  repoLink: string;
```
Comment (SpencerKaiser) [inline packages/database/src/entities/Project.ts:27]: > _In everyday speech, “URL” and “link” are often used interchangeably. Strictly speaking, however, they are different – a link describes the function, which is to take you from one place on the internet to another, while the URL describes the location itself, the place you want to go_

AC mentioned URL

```suggestion
  repoUrl: string;
```

File: packages/database/src/migrations/.snapshot-hangar.json
Comment (SpencerKaiser) [inline packages/database/src/migrations/.snapshot-hangar.json:1]: You'll have to recreate the migration after the changes above

File: packages/shared/src/schema/project/post.ts [hunk 5]
```diff
@@ -16,4 +16,8 @@ export const post = z.object({
     .min(PostValidation.MIN_DESCRIPTION_LENGTH)
     .max(PostValidation.MAX_DESCRIPTION_LENGTH),
   location: z.string().trim().max(PostValidation.MAX_LOCATION_LENGTH).optional(),
+  repoLink: z
```
Comment (SpencerKaiser) [inline packages/shared/src/schema/project/post.ts:19]: Please include test changes to address this addition

File: packages/api/src/api/project/post.ts
Comment (SpencerKaiser) [inline packages/api/src/api/project/post.ts:1]: Tests need to be added to address these changes

**Review by SpencerKaiser — CHANGES_REQUESTED** [review]

File: packages/database/src/entities/Project.ts (hunk 3 above)
Comment (granthill7) [reply packages/database/src/entities/Project.ts:26]: Do we not need to leave the option open for a project to not have a repoLink right off of the bat? This was my call

File: packages/api/src/api/project/post.ts (hunk 2 above)
Comment (granthill7) [reply packages/api/src/api/project/post.ts:31]: So even inside of the try block we should just immediately send the response instead of throwing to the catch block?

File: packages/shared/src/schema/project/post.ts [hunk 6]
```diff
@@ -16,4 +16,8 @@
...
+    .string()
+    .includes('github.com' || 'bitbucket.com' || 'gitlabs.com')
```
[the remark this answers is not in the document]
Comment (SpencerKaiser) [reply packages/shared/src/schema/project/post.ts:21]: @Ansaar-Syed @granthill7 please address the above ^ there might be a way to check that the URL starts with `https://URL`

File: packages/api/src/api/project/post.ts (hunk 2 above)
Comment (SpencerKaiser) [reply packages/api/src/api/project/post.ts:31]: Oh I totally missed that... ok the way yall are doing it is fine but the error message is way too playful for a server; can you sub mine for what you have?

File: packages/api/src/api/project/post.ts (hunk 1 above)
Comment (granthill7) [reply packages/api/src/api/project/post.ts:29]: A try inside of a try?

File: packages/database/src/entities/Project.ts (hunk 3 above)
Comment (SpencerKaiser) [reply packages/database/src/entities/Project.ts:26]: Right now we have no `PUT` endpoint so the project creation needs to be final

File: packages/api/src/api/project/post.ts (hunk 1 above)
Comment (SpencerKaiser) [reply packages/api/src/api/project/post.ts:29]: Yep good call on the try; the name of the variable still needs to change though

File: packages/shared/src/schema/project/post.ts [hunk 7]
```diff
@@ -16,4 +16,8 @@
     .min(PostValidation.MIN_DESCRIPTION_LENGTH)
     .max(PostValidation.MAX_DESCRIPTION_LENGTH),
   location: z.string().trim().max(PostValidation.MAX_LOCATION_LENGTH).optional(),
+  repoUrl: z
+    .string()
+    .includes('github.com' || 'bitbucket.com' || 'gitlabs.com')
```
[the remark this answers is not in the document]
Comment (SpencerKaiser) [reply packages/shared/src/schema/project/post.ts:21]: This is still valid; please review and ask questions if it's unclear what the expectation is

**Review by SpencerKaiser — CHANGES_REQUESTED** [review]
Several comments haven't been addressed yet

File: packages/shared/tests/schema/project/post.test.ts [hunk 8]
```diff
@@ -76,4 +77,8 @@ describe('project post schema', () => {
       }).success,
     ).toBe(false);
   });
+
+  it('validates a matching RepoUrl', () => {
+    expect(Schema.project.post.safeParse(validProject.repoURl).success).toBe(true);
```
Comment (SpencerKaiser) [inline packages/shared/tests/schema/project/post.test.ts:82]: You need to pass a full object to this; look at the tests above for examples. e.g., if you want to override the `description` from the `validProject`, this is what you'd do:
``` typescript
{
    ...validProject,
    description: Array(PostValidation.MIN_DESCRIPTION_LENGTH - 1).fill('A'),
  }
```
you'll need to do something similar and then pass in a good repo url and a bad one and make sure both work as expected

File: packages/shared/src/schema/project/post.ts [hunk 9]
```diff
@@ -16,4 +16,15 @@ export const post = z.object({
...
     .max(PostValidation.MAX_DESCRIPTION_LENGTH),
   location: z.string().trim().max(PostValidation.MAX_LOCATION_LENGTH).optional(),
+  repoUrl: z
+    .string()
+    .trim()
+    .url()
+    .refine(
+      (url) =>
+        url.startsWith('https://github.com/') ||
+        url.startsWith('https://bitbucket.com/') ||
```
Comment (SpencerKaiser) [inline packages/shared/src/schema/project/post.ts:26]: After some research I think it's best we exclude this as an option
```suggestion
```

**Review by SpencerKaiser — CHANGES_REQUESTED** [review]

**Review by SpencerKaiser — DISMISSED** [review]

File: packages/shared/src/schema/project/post.ts [hunk 10]
```diff
@@ -16,4 +16,14 @@ export const post = z.object({
...
     .max(PostValidation.MAX_DESCRIPTION_LENGTH),
   location: z.string().trim().max(PostValidation.MAX_LOCATION_LENGTH).optional(),
+  repoUrl: z
+    .string()
+    .trim()
+    .url()
+    .refine(
+      (url) =>
+        url.startsWith('https://github.com/') ||
+        url.startsWith('https://gitlabs.com/'),
```
Comment (SpencerKaiser) [inline packages/shared/src/schema/project/post.ts:26]:
```suggestion
      (url) => url.startsWith('https://github.com/') || url.startsWith('https://gitlabs.com/'),
```

**Review by SpencerKaiser — APPROVED** [review]
